crates/ark/src/lsp/completions/sources/unique.rs

```diff
@@ -15,61 +15,62 @@ mod string;
-use extractor::completions_from_dollar;
-use namespace::completions_from_namespace;
-use string::completions_from_string;
 use tower_lsp::lsp_types::CompletionItem;
 
-use crate::lsp::completions::parameter_hints::ParameterHints;
-use crate::lsp::document_context::DocumentContext;
+use crate::lsp::completions::builder::CompletionBuilder;
+use crate::lsp::completions::sources::unique::colon::SingleColonSource;
+use crate::lsp::completions::sources::unique::comment::CommentSource;
+use crate::lsp::completions::sources::unique::custom::CustomSource;
+use crate::lsp::completions::sources::unique::extractor::AtSource;
+use crate::lsp::completions::sources::unique::extractor::DollarSource;
+use crate::lsp::completions::sources::unique::namespace::NamespaceSource;
+use crate::lsp::completions::sources::unique::string::StringSource;
+use crate::lsp::completions::sources::CompletionSource;
 
-pub fn completions_from_unique_sources(
-    context: &DocumentContext,
-    parameter_hints: ParameterHints,
-) -> Result<Option<Vec<CompletionItem>>> {
-    log::info!("completions_from_unique_sources()");
+/// Aggregator for unique completion sources
+/// This source tries each unique source in order and returns the first time
+/// a source returns completions (with the caveat that single colon completions
+/// are special).
+pub struct UniqueCompletionsSource;
 
-    // Try to detect a single colon first, which is a special case where we
-    // don't provide any completions
-    if let Some(completions) = completions_from_single_colon(context) {
-        return Ok(Some(completions));
+impl CompletionSource for UniqueCompletionsSource {
+    fn name(&self) -> &'static str {
+        "unique_sources"
     }
 
-    // Try comment / roxygen2 completions
-    if let Some(completions) = completions_from_comment(context)? {
-        return Ok(Some(completions));
-    }
+    fn provide_completions(
+        &self,
+        builder: &CompletionBuilder,
+    ) -> Result<Option<Vec<CompletionItem>>> {
+        log::info!("Getting completions from unique sources");
 
-    // Try string (like file path) completions
-    if let Some(completions) = completions_from_string(context)? {
-        return Ok(Some(completions));
-    }
+        let sources: &[&dyn CompletionSource] = &[
+            // Try to detect a single colon first, which is a special case where we
+            // don't provide any completions
+            &SingleColonSource,
+            &CommentSource,
+            &StringSource,
+            &NamespaceSource,
+            &CustomSource,
+            &DollarSource,
+            &AtSource,
+        ];
 
-    // Try `package::prefix` (or `:::`) namespace completions
-    if let Some(completions) = completions_from_namespace(context, parameter_hints)? {
-        return Ok(Some(completions));
-    }
+        for source in sources {
+            let source_name = source.name();
+            log::debug!("Trying completions from source: {}", source_name);
```

Comment:
Switched to a loop here in unique completions as well.

Reply:
Removed these loops because they come with dynamic dispatch, which we'd rather avoid.